runner.py

```diff
@@ -0,0 +1,252 @@
+    ) as response:
+        token = await response.text()
+        return "Bearer " + token
+
+
+async def ghcr_auth_handler(session: ah.ClientSession, _api_base: str, repo: str) -> str:
+    logging.info("Asked for github token for repo: %s", repo)
+    scope = ":".join(["repo", repo, "pull"])
+    url = "https://ghcr.io/token?" + urllib.parse.urlencode({"scope": scope})
+    async with session.get(url, raise_for_status=True) as response:
+        body = await response.json()
+        token: str = body["token"]
+        return "Bearer " + token
+
+
+async def default_auth_handler(_session: ah.ClientSession, _api_base: str, _repo: str) -> str:
+    token: str = settings.docker_auth.token
+    return "Bearer " + token
+
+
+async def auth(session: ah.ClientSession, api_base: str, repo: str) -> str:
+    if api_base in API_URLs:
+        # Callers are expected to delete items from AUTH_TOKENS when they get a 403.
+        return AUTH_TOKENS[api_base]
+
+    auth_handlers = {
+        "https://registry.hub.docker.com/v2": docker_hub_auth,
+        "https://ghcr.io/v2": ghcr_auth_handler,
+    }
+
+    handler = auth_handlers.get(api_base, default_auth_handler)
+
+    token = await handler(session, api_base, repo)
+    AUTH_TOKENS[api_base] = token
+    return token
+
+
+def _parse_image_ref(image_ref: str) -> urllib.parse.ParseResult:
+    user_url = urllib.parse.urlparse(image_ref)
+
+    # The built-in urllib.parse seems to not understand no-scheme URLs.
+    if not user_url.scheme:
+        user_url = urllib.parse.urlparse("https://" + image_ref)
+
+    return user_url
+
+
+def _get_api_base(image_ref: str) -> str:
+    user_url = _parse_image_ref(image_ref)
+
+    # TODO: Find a way to do this that doesn't rely on lookup tables
+    api_base = API_URLs.get(user_url.netloc, None)
+    if api_base is None:
+        raise ValueError(f"Registry not supported: {image_ref}")
+
+    return api_base
+
+
+async def get_remote_tags(
+    session: ah.ClientSession, image_ref: str, auth_token: Optional[str] = None
+) -> list[str]:
+    """Use the OCI Distribution API to get a list of tags for an image."""
+
+    user_url = _parse_image_ref(image_ref)
+    api_base = _get_api_base(image_ref)
+    repo = user_url.path.strip("/")
+
+    target_url = "/".join([api_base, user_url.path, "tags/list"])
+
+    auth_token = auth_token or await auth(session, api_base, repo)
+    headers = {
+        "Authorization": auth_token,
+        "Accept": "application/vnd.docker.distribution.manifest.v2+json",
+    }
+
+    async with session.get(target_url, headers=headers, raise_for_status=True) as response:
+        # FIXME: We should be expecting at least 404 & 403 here.
+        # TODO: This is actually a paginated API. Need `async yield` + multiple requests.
+        # Yeah, the typechecker has no way of confirming what's up.
+        # The remote API could change at any time. Such is life.
+        rv: list[str] = (await response.json())["tags"]
```

Comment:
while its true typeck cant look into `Dict[str, Any]` soup, we can add validation in the form of typeguarding and just checking it is the type we expect, something to get to later though